README.md
# Middager
Displaying a google-kalender on a 8x8 diode matrix display
Using the library to drive the dot matrix led display cascaded:
https://pypi.org/project/luma.led_matrix/


# Requirements:
```
pip3 install --upgrade luma.led_matrix
pip3 install --upgrade google-api-python-client google-auth-httplib2 google-auth-oauthlib
pip3 install --upgrade python-dateutil
pip3 install backports-datetime-fromisoformat
```
In case of error message on installing the libraries, this might work:
```
sudo apt-get install libjpeg-dev zlib1g-dev
pip install Pillow
```

# Installation
Make sure the SPI is enabled.
- Go to https://developers.google.com/calendar/quickstart/python
- run the "Enable the Google Calendar API". 
- Give the project a name and hit "Next"
- Select "Desktop app" and hit "Create"
- Press `Download Client Configuration` and save in root directory (/home/pi/) as ".creds.json". Full pat should be `/home/pi/.creds.json`

Connect the display to the serial ports pins with Raspi IO pins

Run the program using command:
```
python3 Middager/middag.py
```
When powering up, have the led display disconnected from power as the load might make starting up the Pi a strain on the input voltage.

In case of error message from the dot-matrix driver, try chaning line 69 of middag.py to
```
        serial = spi(port=0, device=0, cs_high=True, gpio=noop())
```
## Command line arguments
The program iterprets some command line arguments. Tese are:
|Argument|Description|
|----------|----------|
|-h, --help| The help instructions|
|-m, --mode| Repeat mode. Specify 'loop' as argument if the program should enter an eternal loop |

Specifying nothing will keep the program running for 10 minutes approximately, suitable for a cron based run schedule
# Setting up continuous running from boot
## CRON run
To run the display continuously, enter a crontab line like this:
```
*/10 * * * * python3 /home/pi/Middager/middag.py
```
## Service
Alternatively, install as service by copying the service file `/lib/systemd/system` using root:
```
sudo cp  middager-sync.service /lib/systemd/system/middager-sync.service
```
When service file is in place, run the following commands to verify and start the service:
```
sudo chmod 644 /lib/systemd/system/middager-sync.service
chmod +x /home/pi/Middager/middager-sync.service
sudo systemctl daemon-reload
sudo systemctl enable middager-sync.service
sudo systemctl start middager-sync.service
```


CHANGELOG.md
# Changelog

## [1.2.2](https://github.com/stenjo/Middager/compare/v1.2.1...v1.2.2) (2023-11-07)


### Bug Fixes

* **dateHandling:** trouble with dates from appointment ([3da38d5](https://github.com/stenjo/Middager/commit/3da38d5e796ae35e848568e905a1eafc22e9983f))

## [1.2.1](https://github.com/stenjo/Middager/compare/v1.2.0...v1.2.1) (2023-11-04)


### Bug Fixes

* **date handling:** can't subtract offset-naive and offset-aware datetimes ([025f8af](https://github.com/stenjo/Middager/commit/025f8af79bca06104c6c3ca66df5445e003dc85f))

## [1.2.0](https://github.com/stenjo/Middager/compare/v1.1.0...v1.2.0) (2023-10-28)


### Features

* vis dato for hendelser mer enn 1 uke fram ([a0939cc](https://github.com/stenjo/Middager/commit/a0939ccd017f77d5a739c863539fdb99855c36cb))

## [1.1.0](https://github.com/stenjo/Middager/compare/v1.0.1...v1.1.0) (2022-10-03)


### Features

* display version on startup ([f4c12ef](https://github.com/stenjo/Middager/commit/f4c12efd16baeedb2cfa83f0f3f8bb0c7cd7098b))
* switch off betw. 23_30 and 6:45 ([dbee752](https://github.com/stenjo/Middager/commit/dbee7524ef15cb920c17bee0cea60e5d95a71136))

## [1.0.1](https://github.com/stenjo/Middager/compare/v1.0.0...v1.0.1) (2022-09-29)


### Bug Fixes

* handle api error on calendar fetch ([3f05256](https://github.com/stenjo/Middager/commit/3f052563afe7cef1d080b80a2ebfb285287afbbf))

## 1.0.0 (2022-08-21)


### Features

* add command line option for continuous loop ([36b677a](https://github.com/stenjo/Middager/commit/36b677a61f930e44b53430e63fcdcd8ebc03dd66))
* add service option to continous running carousel ([513fb27](https://github.com/stenjo/Middager/commit/513fb279ce4a9ba8f9194607108d41411c11535f))
* add servise handle as option to cron tab ([f3aa4c0](https://github.com/stenjo/Middager/commit/f3aa4c0c8ed95600f66cced4e05edf40ecfd008a))


### Bug Fixes

* addin eternal running param to execution ([c917ae8](https://github.com/stenjo/Middager/commit/c917ae880cad5cbff271b5df11fdb81a40279cc8))
* running service from Middager ([733e57e](https://github.com/stenjo/Middager/commit/733e57ecfa239fccafc4550884f94cd2264d481b))
* update documentation service parameters ([8c132e1](https://github.com/stenjo/Middager/commit/8c132e100b5394c77f719558cc759f5dd45a3bfa))
* update exec path in service file ([a50d785](https://github.com/stenjo/Middager/commit/a50d785d51017964d764bb6b4e580b25f237c679))
* update README file with service instructions ([df77116](https://github.com/stenjo/Middager/commit/df7711611a07b693e6d9ab23678152a4c9fe9526))
* update service file fixing access and working directories ([5683173](https://github.com/stenjo/Middager/commit/5683173d3a4b6955b7b62841edafb60e71274f70))
* using default python ([bdd7929](https://github.com/stenjo/Middager/commit/bdd7929c47478677fe0fd042d1056739b383e2c3))


tests/dateHandling_test.py
from datetime import datetime, timezone
from dateHandling import dayText, isNowInTimePeriod

def test_isNowInTimePeriod():
    startTime = datetime.strptime("2023-10-22T10:12", "%Y-%m-%dT%H:%M")
    endTime = datetime.strptime("2023-10-25T10:12", "%Y-%m-%dT%H:%M")
    today = datetime.strptime("2023-10-23T10:12", "%Y-%m-%dT%H:%M")
    before = datetime.strptime("2023-10-22T10:11", "%Y-%m-%dT%H:%M")
    after = datetime.strptime("2023-10-25T10:13", "%Y-%m-%dT%H:%M")
    
    assert isNowInTimePeriod(startTime, endTime, today) == True
    assert isNowInTimePeriod(startTime, endTime, before) == False
    assert isNowInTimePeriod(startTime, endTime, after) == False


today = datetime.strptime("2023-10-23T10:12", "%Y-%m-%dT%H:%M").replace(tzinfo=timezone.utc)
    
def test_dayText_tomorrow():
    event = {
        'start': {
            'date': '2023-10-24'
        },
        'summary': 'Event summary'
    }
    
    assert dayText(event, today) == "I morgen: Event summary"

def test_dayText_today():
    event = {
        'start': {
            'date': '2023-10-23'
        },
        'summary': 'Event summary'
    }
    
    assert dayText(event, today) == "I dag: Event summary"
    
def test_dayText_today_with_time():
    event = {
        'start': {
            'date': '2023-10-23',
            'dateTime': '2023-10-23T10:22:18'
        },
        'summary': 'Event summary'
    }
    
    assert dayText(event, today) == "I dag: Event summary kl. 10:22"
    
def test_dayText_friday():
    event = {
        'start': {
            'date': '2023-10-27'
        },
        'summary': 'Event summary'
    }
    
    assert dayText(event, today) == "Fredag: Event summary"
    
def test_dayText_friday_plus_one_week():
    event = {
        'start': {
            'date': '2023-11-03'
        },
        'summary': 'Event summary'
    }
    
    assert dayText(event, today) == "Fredag 03/11: Event summary"

def test_dayText_today_with_time_and_timezone():
    event = {
        'start': {
            'date': '2023-10-23',
            'dateTime': '2023-10-23T10:22:18+01:00'
        },
        'summary': 'Event summary'
    }
    today = datetime.strptime("2023-10-23T10:12 +0100", "%Y-%m-%dT%H:%M %z")
    # print(today)
    # dt:  2023-11-09 13:15:00+01:00
    # today:  2023-11-04 18:54:37.512734

    assert dayText(event, today) == "I dag: Event summary kl. 10:22"
    
def test_dayText_today_with_datetime_and_timezone():
    event = {
        'start': {
            'date': datetime.today().strftime('%Y-%m-%d'),
            'dateTime': datetime.today().strftime('%Y-%m-%d') + 'T10:22:18+01:00'
        },
        'summary': 'Event summary'
    }

    assert dayText(event) == "I dag: Event summary kl. 10:22"

def test_dayText_today_with_date_and_no_timezone():
    event = {
        'start': {
            'date': datetime.today().strftime('%Y-%m-%d'),
        },
        'summary': 'Event summary'
    }

    assert dayText(event) == "I dag: Event summary"



dateHandling.py
from datetime import datetime, timezone, timedelta

def isNowInTimePeriod(startTime, endTime, nowTime): 
    if startTime < endTime: 
        return nowTime >= startTime and nowTime <= endTime 
    else: 
        #Over midnight: 
        return nowTime >= startTime or nowTime <= endTime 

def dayText(event, today=datetime.now(timezone.utc)):

    weekday = ['Mandag', 'Tirsdag', 'Onsdag', 'Torsdag', 'Fredag','Lørdag','Søndag']
    text = ''
    # The line `dt = datetime.datetime.fromisoformat(event['start'].get('dateTime',
    # event['start'].get('date')))`
    # is parsing the start time of an event from a dictionary called `event`.
    dt = datetime.fromisoformat(event['start'].get('dateTime', event['start'].get('date')))
    tomorrow = today + timedelta(1)
    delta = dt - today

    if dt.date() == today.date() :
        text = text + 'I dag: '
    elif dt.date() == tomorrow.date() :
        text = text + 'I morgen: '
    elif delta.days > 6:
        text = text + weekday[dt.weekday()] + ' ' + dt.strftime('%d/%m') + ': '
    else :
        text = text + weekday[dt.weekday()] + ': '

    text = text + event['summary']

    if dt.hour > 0:
        text = text + dt.strftime(' kl. %H:%M')

    return text


middag.py
from __future__ import print_function
import datetime
from re import T
import time
import pickle
import os.path
import sys, getopt

from googleapiclient.discovery import build
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
from luma.led_matrix.device import max7219
from luma.core.interface.serial import spi, noop
from luma.core.render import canvas
from luma.core.virtual import viewport
from luma.core.legacy import text, show_message
from luma.core.legacy.font import proportional, CP437_FONT, TINY_FONT, SINCLAIR_FONT, LCD_FONT
from backports.datetime_fromisoformat import MonkeyPatch
from pathlib import Path
from dateHandling import isNowInTimePeriod, dayText

MonkeyPatch.patch_fromisoformat()
# If modifying these scopes, delete the file token.pickle.
SCOPES = ['https://www.googleapis.com/auth/calendar.readonly']


def main(argv):
    loop = None
    try:
        opts, args = getopt.getopt(argv,"hm:",["help ","mode ="])
    except getopt.GetoptError:
        print('middag.py -m <loop>')
        sys.exit(2)
    for opt, arg in opts:
        if opt in ('-h', '--help'):
            print('middag.py -m <loop>')
            sys.exit()
        elif opt in ("-m", "--mode"):
            if arg in ('loop'):
                loop = True

    print (loop)
    """Shows basic usage of the Google Calendar API.
    Prints the start and name of the next 10 events on the user's calendar.
    """

    # Get version
    script_dir = Path( __file__ ).parent.absolute()
    filename = os.path.join( script_dir,'version.txt')
    version = Path(filename).read_text().replace('\n','')
    print("Version: ", version)
    
    creds = None
    # The file token.pickle stores the user's access and refresh tokens, and is
    # created automatically when the authorization flow completes for the first
    # time.
    if os.path.exists('token.pickle'):
        with open('token.pickle', 'rb') as token:
            creds = pickle.load(token)
    # If there are no (valid) credentials available, let the user log in.
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                '/home/pi/.creds.json', SCOPES)
            creds = flow.run_local_server()
        # Save the credentials for the next run
        with open('token.pickle', 'wb') as token:
            pickle.dump(creds, token)

    service = build('calendar', 'v3', credentials=creds)

    # create matrix device
    serial = spi(port=0, device=0, gpio=noop())
    device = max7219(serial, cascaded=8, block_orientation=90, rotate=0)
    print("Created device")
    
    # Show version
    show_message(device, "Ver. " + version, fill="white", font=proportional(CP437_FONT))
    time.sleep(4)
    
    start = time.perf_counter()
    while (start + 550) > time.perf_counter() or loop:
        
        if not isNowInTimePeriod(datetime.time(6,45), datetime.time(23,30), datetime.datetime.now().time()):
            time.sleep(300)
            continue
        
        # Call the Calendar API
        now = datetime.datetime.utcnow().isoformat() + 'Z' # 'Z' indicates UTC time
        # print('Getting the upcoming 10 events')
        try:
            events_result = service.events().list(calendarId='parterapeutene.no_e1or90m2lp6p523ma7u15v2pc0@group.calendar.google.com', timeMin=now,
                                                maxResults=7, singleEvents=True,
                                                orderBy='startTime').execute()
        except:
            time.sleep(2)
            continue
            
        events = events_result.get('items', [])

        msg = ''

        if not events:
            msg = '- finner ingen middager - '
        for event in events:

            msg = dayText(event)
            show_message(device, msg, fill="white", font=proportional(CP437_FONT))
            time.sleep(2)

if __name__ == '__main__':
    main(sys.argv[1:])
